File: apps/client-macos/Sources/TimeTrack/Sync/ScreenshotSyncEngine.swift
import Foundation

/// PRD §7.4/§7.5 — one-way (client→server) drain of the durable image buffer. Mirrors `SyncEngine`
/// but for a single binary kind and its own multipart uploader (2.2a mandates a separate path).
/// Each image is removed on confirmed success (201), dropped on a permanent 4xx (poison can't wedge
/// the queue), or kept on transient/auth (caller backs off). Stale images age/count-prune each cycle.
/// Not a capture path → not AckGate-gated; safe on both the online and offline-marker branches.
final class ScreenshotSyncEngine {
    private let buffer: ImageBufferStore
    private let uploader: ScreenshotUploading
    private let backoff: BackoffPolicy
    private let intervalSeconds: TimeInterval
    private let batchLimit: Int
    private let maxAge: TimeInterval
    private let maxCount: Int

    private var timer: Timer?
    private var started = false
    private var isDraining = false

    init(buffer: ImageBufferStore, uploader: ScreenshotUploading,
         backoff: BackoffPolicy = BackoffPolicy(),
         intervalSeconds: TimeInterval = 90, batchLimit: Int = 20,
         maxAge: TimeInterval = 7 * 24 * 3600, maxCount: Int = 500) {
        self.buffer = buffer
        self.uploader = uploader
        self.backoff = backoff
        self.intervalSeconds = intervalSeconds
        self.batchLimit = batchLimit
        self.maxAge = maxAge
        self.maxCount = maxCount
    }

    func start() {
        guard !started else { return }
        started = true
        scheduleNext(after: 0)
    }

    func stop() {
        started = false
        timer?.invalidate()
        timer = nil
    }

    /// One drain pass. Returns true if it stopped early on a transient/auth failure.
    @discardableResult
    func syncNow() async -> Bool {
        guard !isDraining else { return false }
        isDraining = true
        defer { isDraining = false }

        buffer.prune(olderThan: maxAge, maxCount: maxCount)

        for record in buffer.take(limit: batchLimit) {
            guard let jpeg = try? Data(contentsOf: record.url) else {
                buffer.remove(id: record.id)   // unreadable file can't be uploaded — drop it
                continue
            }
            switch await uploader.upload(id: record.id, capturedAt: record.capturedAt,
                                         group: record.group, jpeg: jpeg) {
            case .success:
                buffer.remove(id: record.id)
                backoff.reset()
            case .permanent:
                buffer.remove(id: record.id)
            case .transient, .authFailed:
                return true
            }
        }
        return false
    }

    // MARK: - self-scheduling timer glue (build-verified)

    private func runCycle() async {
        let backedOff = await syncNow()
        guard started else { return }
        scheduleNext(after: backedOff ? backoff.nextDelay() : intervalSeconds)
    }

    private func scheduleNext(after delay: TimeInterval) {
        timer?.invalidate()
        let t = Timer(timeInterval: max(0.001, delay), repeats: false) { [weak self] _ in
            Task { await self?.runCycle() }
        }
        RunLoop.main.add(t, forMode: .common)
        timer = t
    }
}

File: apps/client-windows/src/NiftyTimer/Sync/ScreenshotSyncEngine.cs
using NiftyTimer.Storage;

namespace NiftyTimer.Sync;

/// <summary>
/// PRD §7.4 / §7.5 — one-way (client → server) drain of the durable image buffer.
///
/// Mirrors <see cref="SyncEngine"/> but carries a single binary kind through its own multipart
/// uploader. Each image is removed on confirmed success, dropped on a permanent 4xx so a poison
/// record cannot wedge the queue, or kept on a transient/auth failure — in which case the cycle
/// stops early and the caller backs off. Stale images age- and count-prune each cycle.
///
/// **The local file is deleted only after a confirmed upload** (PRD §6.2). Removing it optimistically
/// on send would lose the capture whenever the response is the thing that goes missing.
///
/// Not a capture path, so it is not gated and runs on both the online and the offline-marker
/// branches — draining what was already captured is not capturing.
/// </summary>
public sealed class ScreenshotSyncEngine : IDisposable
{
    private readonly IImageBuffer _buffer;
    private readonly IScreenshotUploading _uploader;
    private readonly BackoffPolicy _backoff;
    private readonly TimeSpan _interval;
    private readonly int _batchLimit;
    private readonly TimeSpan _maxAge;
    private readonly int _maxCount;
    private readonly Lock _gate = new();
    private readonly SemaphoreSlim _drainGate = new(1, 1);

    private Timer? _timer;
    private bool _started;
    private bool _disposed;

    public ScreenshotSyncEngine(
        IImageBuffer buffer,
        IScreenshotUploading uploader,
        BackoffPolicy? backoff = null,
        TimeSpan? interval = null,
        int batchLimit = 20,
        TimeSpan? maxAge = null,
        int maxCount = 500)
    {
        _buffer = buffer;
        _uploader = uploader;
        _backoff = backoff ?? new BackoffPolicy();
        _interval = interval ?? TimeSpan.FromSeconds(90);
        _batchLimit = batchLimit;
        _maxAge = maxAge ?? TimeSpan.FromDays(7);
        _maxCount = maxCount;
    }

    public void Start()
    {
        lock (_gate)
        {
            if (_started)
            {
                return;
            }

            _started = true;
        }

        ScheduleNext(TimeSpan.Zero);
    }

    public void Stop()
    {
        lock (_gate)
        {
            _started = false;
            _timer?.Dispose();
            _timer = null;
        }
    }

    /// <summary>One drain pass, skipped if another is already running.</summary>
    public async Task<bool> SyncNowAsync(CancellationToken cancellationToken = default)
    {
        if (!await _drainGate.WaitAsync(0, cancellationToken).ConfigureAwait(false))
        {
            return false;
        }

        try
        {
            return await DrainAsync(cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            _drainGate.Release();
        }
    }

    /// <summary>
    /// One drain pass that WAITS for any in-flight cycle rather than skipping. Sign-out must call
    /// this: <see cref="SyncNowAsync"/> returns immediately when a cycle is running, so using it
    /// as the final drain means the drain silently does not happen and the buffer is then cleared
    /// anyway — discarding captures that were never sent.
    /// </summary>
    public async Task<bool> FlushAsync(CancellationToken cancellationToken = default)
    {
        await _drainGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            return await DrainAsync(cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            _drainGate.Release();
        }
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        Stop();
        _drainGate.Dispose();
    }

    private async Task<bool> DrainAsync(CancellationToken cancellationToken)
    {
        _buffer.Prune(_maxAge, _maxCount);

        foreach (var record in _buffer.Take(_batchLimit))
        {
            cancellationToken.ThrowIfCancellationRequested();

            byte[] jpeg;
            try
            {
                jpeg = File.ReadAllBytes(record.Path);
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException)
            {
                // An image we cannot read can never be uploaded, and retrying it forever would
                // block every capture behind it. Drop it.
                _buffer.Remove(record.Id);
                continue;
            }

            var result = await _uploader
                .UploadAsync(record.Id, record.CapturedAt, record.Group, jpeg, cancellationToken)
                .ConfigureAwait(false);

            switch (result)
            {
                case UploadResult.Success:
                    _buffer.Remove(record.Id);
                    _backoff.Reset();
                    break;
                case UploadResult.Permanent:
                    _buffer.Remove(record.Id); // poison record — drop it rather than wedge the queue
                    break;
                default:
                    return true; // transient / auth — stop this cycle, caller backs off
            }
        }

        return false;
    }

    private async void RunCycleAsync()
    {
        bool backedOff;
        try
        {
            backedOff = await SyncNowAsync().ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            return;
        }

        lock (_gate)
        {
            if (!_started)
            {
                return;
            }
        }

        ScheduleNext(backedOff ? _backoff.NextDelay() : _interval);
    }

    private void ScheduleNext(TimeSpan delay)
    {
        lock (_gate)
        {
            if (!_started)
            {
                return;
            }

            _timer?.Dispose();
            _timer = new Timer(
                static state => ((ScreenshotSyncEngine)state!).RunCycleAsync(),
                this,
                delay < TimeSpan.Zero ? TimeSpan.Zero : delay,
                Timeout.InfiniteTimeSpan);
        }
    }
}
